graphbolt/include/graphbolt/rowwise_pick.h

```diff
@@ -0,0 +1,184 @@
+
+/**
+ * @brief Performs row-wise picking based on the given parameters.
+ *
+ * @param graph The pointer to a csc sampling graph.
+ * @param rows The tensor containing the row indices.
+ * @param num_picks The tensor containing the number of picks per edge type.
+ * @param probs Optional tensor containing probabilities for picking.
+ * @param return_eids Boolean indicating if edge IDs need to be returned. The
+ * last TensorList in the tuple is this value when required.
+ * @param replace Boolean indicating if picking is done with replacement.
+ * @param pick_fn The function used for picking.
+ * @return A tuple containing the picked rows, picked columns, picked etypes,
+ * and picked edge IDs (if required).
+ */
+std::tuple<torch::Tensor, torch::Tensor, torch::Tensor, torch::Tensor>
+RowWisePickPerEtype(
+    const CSCPtr graph, const torch::Tensor& rows,
+    const torch::Tensor& num_picks, const torch::optional<torch::Tensor>& probs,
+    bool return_eids, bool replace, RangePickFn pick_fn) {
+  const auto indptr = graph->CSCIndptr();
+  const auto indices = graph->Indices();
+  const auto type_per_edge = graph->TypePerEdge().value();
+  const int64_t num_rows = rows.size(0);
+  const int64_t num_etypes = num_picks.size(0);
+  std::vector<int64_t> picked_num_per_row(num_rows + 1);
+  TensorList picked_cols_per_row(
+      num_rows, torch::tensor({}, indices.options()));
+  TensorList picked_etypes_per_row(
+      num_rows, torch::tensor({}, type_per_edge.options()));
+  TensorList picked_eids_per_row;
+  if (return_eids) {
+    picked_eids_per_row.resize(num_rows, torch::tensor({}, indptr.options()));
+  }
+  int64_t min_num_pick = -1;
+  auto num_picks_accessor = num_picks.accessor<int64_t, 1>();
+  bool pick_all = true;
+  for (int i = 0; i < num_etypes; i++) {
+    int64_t num_pick = num_picks_accessor[i];
+    if (num_pick != -1) {
+      if (min_num_pick == -1 || num_pick < min_num_pick)
+        min_num_pick = num_pick;
+      pick_all = false;
+    }
+  }
+
+  auto type_per_edge_accessor = type_per_edge.accessor<int64_t, 1>();
+  const bool has_probs = probs.has_value();
+  auto thread_num = compute_num_threads(0, num_rows, kDefaultPickGrainSize);
+  std::vector<size_t> block_offset(thread_num + 1, 0);
+
+  torch::parallel_for(
+      0, num_rows, kDefaultPickGrainSize, [&](size_t b, size_t e) {
+        int64_t sampled_num_this_thread = 0;
+        for (size_t i = b; i < e; ++i) {
+          const auto rid = rows[i].item<int64_t>();
+          TORCH_CHECK(rid >= 0 && rid < graph->NumNodes());
+          const auto off = indptr[rid].item<int64_t>();
+          const auto len = indptr[rid + 1].item<int64_t>() - off;
+
+          if (len == 0) continue;
+
+          torch::Tensor picked_indices_this_row;
+          // Fast path.
+          if ((pick_all || len <= min_num_pick) && !replace) {
+            picked_indices_this_row =
+                torch::arange(off, off + len, indptr.options());
```

Comment:
Is calling PyTorch functions in a `torch::parallel_for` loop a good practice?  I'm not sure, but I feel @jermainewang 's concern of returning a PyTorch tensor within a parallel loop also applies here.